app/src/main/java/com/azuresamples/msalnativeauthandroidkotlinsampleapp/MFAFragment.kt

```diff
@@ -0,0 +1,268 @@
+        updateUI(STATUS.SignedIn)
+        displayAccount(accountState)
+    }
+
+    private fun displaySignedOutState() {
+        emptyFields()
+        updateUI(STATUS.SignedOut)
+        emptyResults()
+    }
+
+    private fun updateUI(status: STATUS) {
+        when (status) {
+            STATUS.SignedIn -> {
+                binding.signIn.isEnabled = false
+                binding.signOut.isEnabled = true
+            }
+            STATUS.SignedOut -> {
+                binding.signIn.isEnabled = true
+                binding.signOut.isEnabled = false
+            }
+        }
+    }
+
+    private fun emptyFields() {
+        binding.emailText.setText("")
+        binding.passwordText.setText("")
+    }
+
+    private fun emptyResults() {
+        binding.resultAccessToken.text = ""
+        binding.resultIdToken.text = ""
+    }
+
+    private fun displayAccount(accountState: AccountState) {
+        CoroutineScope(Dispatchers.Main).launch {
+            val accessTokenResult = accountState.getAccessToken()
+            when (accessTokenResult) {
+                is GetAccessTokenResult.Complete -> {
+                    val accessToken = accessTokenResult.resultValue.accessToken
+                    binding.resultAccessToken.text = getString(R.string.result_access_token_text) + accessToken
+
+                    val idToken = accountState.getIdToken()
+                    binding.resultIdToken.text = getString(R.string.result_id_token_text) + idToken
+                }
+                is GetAccessTokenError -> {
+                    displayDialog(getString(R.string.msal_exception_title), accessTokenResult.exception?.message)
+                }
+            }
+        }
+    }
+
+    private fun handleSignInError(error: SignInError) {
+        when {
+            error.isInvalidCredentials() || error.isBrowserRequired() || error.isUserNotFound() -> {
+                displayDialog(error.error, error.errorMessage)
+            }
+            else -> {
+                // Unexpected error
+                displayDialog(getString(R.string.unexpected_sdk_error_title), error.exception?.message)
+            }
+        }
+    }
+
+    private fun displayDialog(error: String? = null, message: String?) {
+        val builder = AlertDialog.Builder(requireContext())
+        builder.setTitle(error)
+            .setMessage(message)
+        val alertDialog = builder.create()
+        alertDialog.show()
+    }
+
+    private fun displayMFARequiredDialog(actionResult: SignInResult.MFARequired) {
+        val builder = AlertDialog.Builder(requireContext())
+        builder.setTitle(R.string.mfa_required_notice)
+
+        // If proceed
+        builder.setPositiveButton(getString(R.string.yes_message)) { dialog, which ->
+            CoroutineScope(Dispatchers.Main).launch {
+                val awaitingMFAState = actionResult.nextState
+                val requestDefaultAuthMethodResult = awaitingMFAState.requestChallenge()
```

Comment:
nit: while technically this variable name is correct, I'm not sure we should draw attention to the difference between "default auth methods" vs. "non default auth methods", as we're only showcasing 1 flow (the default one) here. Naming this just `challengeResult` or something similar might be simple enough.